adaptive/priority_queue.py

```diff
@@ -0,0 +1,104 @@
+from sortedcontainers import SortedDict, SortedList
+
+__all__ = ["Empty", "Queue"]
+
+
+class Empty(KeyError):
+    pass
+
+
+class Queue:
+    """Priority queue supporting update and removal at arbitrary position.
+
+    Parameters
+    ----------
+    entries : iterable of (item, priority)
+        The initial data in the queue. Providing this is faster than
+        calling 'insert' a bunch of times.
+    """
+
+    def __init__(self, entries=()):
+        self._queue = SortedDict(
+            ((priority, -n), item) for n, (item, priority) in enumerate(entries)
+        )
+        # 'self._queue' cannot be keyed only on priority, as there may be several
+        # items that have the same priority. To keep unique elements the key
+        # will be '(priority, self._n)', where 'self._n' is decremented whenever
+        # we add a new element. 'self._n' is negative so that elements with equal
+        # priority are sorted by insertion order.
+        self._n = -len(self._queue)
+        # To efficiently support updating and removing items if their priority
+        # is unknown we have to keep the reverse map of 'self._queue'. Because
+        # items may not be hashable we cannot use a SortedDict, so we use a
+        # SortedList storing '(item, key)'.
+        self._items = SortedList(((v, k) for k, v in self._queue.items()))
+
+    def __len__(self):
+        return len(self._queue)
+
+    def items(self):
+        "Return an iterator over the items in the queue in arbitrary order."
+        return reversed(self._queue.values())
```

Comment:
Why do we use `reversed` here and later if the order is arbitrary?

Reply:
changed the API to return in priority order